Modules/Tokens/src/Tokens.Application/Tokens/Queries/GetToken/Handler.cs

```diff
@@ -29,9 +31,36 @@ private async Task<Token> GetToken(string tokenId, byte[]? password, Cancellatio
         var token = await _tokensRepository.Find(TokenId.Parse(tokenId), _userContext.GetAddressOrNull(), cancellationToken, true) ??
                     throw new NotFoundException(nameof(Token));
 
-        if (!token.CanBeCollectedUsingPassword(_userContext.GetAddressOrNull(), password))
-            throw new NotFoundException(nameof(Token));
+        var identity = _userContext.GetAddressOrNull();
+        if (identity == null)
+        {
+            await EnsureTokenCanBeCollected(token, null, password, cancellationToken);
+        }
+        else if (identity != token.CreatedBy)
+        {
+            if (!token.HasAllocationForIdentity(identity))
+            {
+                await EnsureTokenCanBeCollected(token, identity, password, cancellationToken);
+
+                token.AddAllocationFor(identity, _userContext.GetDeviceId());
+                await _tokensRepository.Update(token, cancellationToken);
+            }
+        }
 
         return token;
     }
+
+    private async Task EnsureTokenCanBeCollected(Token token, IdentityAddress? address, byte[]? password, CancellationToken cancellationToken)
+    {
+        if (token.IsLocked) throw new ApplicationException(ApplicationErrors.TokenIsLocked());
+
+        if (!token.CanBeCollectedUsingPassword(address, password))
+        {
+            token.IncrementAccessFailedCount();
+            await _tokensRepository.Update(token, cancellationToken);
+
+            if (token.IsLocked) throw new ApplicationException(ApplicationErrors.TokenIsLocked());
+            throw new NotFoundException(nameof(Token));
+        }
```

Comment:
I'd suggest to extract more logic into the Token class. For example by doing something like this:

```c#
var error = token.TryToAccess(address, password);

if (error == null)
    return;

if(error == DomainErrors.WrongPassword()){
    await _tokensRepository.Update(token, cancellationToken);
    throw new NotFoundException(nameof(Token));
}

throw new DomainException(error);
```

This was just a quick try. I may have forgotten something or there may be a better way. But the big advantage of this is that the main part of the logic is now in Domain, which makes it a lot easier to test.